README.md
[![](https://ptkdev.it/img/bot/twitter-bot-lib.png)](https://twitter.bot.ptkdev.io)

# Twitter Bot (Library)

[![](https://img.shields.io/badge/license-GLPv3-brightgreen.svg)](#) [![](https://img.shields.io/badge/powered%20by-puppeteer-46aef7.svg)](https://github.com/GoogleChrome/puppeteer) [![](https://img.shields.io/badge/version-v0.2.5-lightgrey.svg)](https://github.com/social-manager-tools/twitter-bot.js/releases) [![](https://img.shields.io/badge/chat%20on-slack-orange.svg)](https://slack.ptkdev.io) [![](https://img.shields.io/badge/blog-medium-2AE176.svg)](http://blog.ptkdev.io) [![](https://img.shields.io/badge/twitter-ptkdevio-2AA3EF.svg)](https://twitter.com/ptkdevio) [![](https://img.shields.io/badge/help-support@ptkdev.io-fbbc05.svg)](mailto:support@ptkdev.io)

[![](https://img.shields.io/badge/donate-patreon-F87668.svg)](http://patreon.ptkdev.io) [![](https://img.shields.io/badge/donate-paypal-46AFE0.svg)](http://paypal.ptkdev.io) [![](https://img.shields.io/badge/buy%20me-coffee-4B788C.svg)](http://coffee.ptkdev.io)

[![](https://ptkdev.it/img/bot/ptkdev-twitter-bot.gif)](https://twitter.bot.ptkdev.io)

## What does it do 
This library (node module) provides api for include [twitter-bot.js](https://github.com/social-manager-tools/twitter-bot.js) function in your application. This bot helps you increase the engagement of your Twitter profile through different social algorithms. Increase the likes on your tweets and followers.

## Features
* [✓] Easy to use
* [✓] Login
* [✓] 2FA (sms pin enabled)
* [✓] Multi-Session
* [✓] Multi-Platform (Windows 10, Mac OSX, Linux, [Raspberry PI 3](https://github.com/social-manager-tools/twitter-bot.js/blob/master/INSTALL.md))
* [✓] Errors manager (bad pin, bad password)
* [✓] Screenshot and Verbose logger
* [✓] Like Mode Realistic: bot select random hashtag from config list and like fast 19-20 tweet, sleep 15-20min and repeat this all time. Sleep at night.
* [✓] Retweet Mode Realistic: bot select random hashtag from config list and rt fast 19-20 tweet, sleep 15-20min and repeat this all time. Sleep at night.
* [✘] Comment Mode Classic: select random comment and random hashtag and write comment under tweet.
* [✘] Follow/Defollow Classic: follow user from random hashtag and defollow after 1h.
* [✘] Defollow All: defollow all your following (ignore users in whitelist).

## Fast usage
1. Run `npm install twitterbotlib`
2. Get [config.js](https://raw.githubusercontent.com/social-manager-tools/twitter-bot-lib/0.2.5/config.js.tpl) file for step 3, fill it properly and remove `.tpl` suffix.
3. On your code require library, config and run bot, example:
```
    const config = require ("./config");
    const Bot = require("twitterbotlib");
    let bot = new Bot(config);
    bot.start();
```
4. If work add star :star: at this project :heart:
5. If you want help me: **[donate on paypal](http://paypal.ptkdev.io)** or become a **[backer on patreon](http://patreon.ptkdev.io)**..

For advanced configuration see [INSTALL.md](https://github.com/social-manager-tools/twitter-bot-lib/blob/master/INSTALL.md).

#### 2FA: SMS Pin
If you received sms or email pin edit `loginpin.txt` and insert it on first line. Wait 50-60 seconds...

#### Tips: hide browser
Edit `config.js` and switch `chrome_headless` option to `true`.

#### Check if work:
See images in ./logs/screenshot or disable `chrome_headless` flag.

## API 
Method | Description
------------ | -------------
start() | run twitter.bot.js

## Desktop setup (GUI Version)
1. Download [Social Manager Tools GUI](https://socialmanagertools.ptkdev.io/).
2. Run application.

## Docker setup
If you prefer to run this using Docker, an official container is available from the [Docker Hub](https://hub.docker.com/r/socialmanagertools/twitter-bot.js).

In order to run it, copy the `config.js.tpl` file, configure it as you prefer, then use it through volume mapping,
like in this example:

```sh
$ docker run \
    --restart=always \
    --name=twitter-bot \
    -d \
    -v /path/to/config.js:/app/config.js \
    socialmanagertools/twitter-bot.js &>/dev/null
```

## Roadmap
See full roadmap (open task, todo and bugs) in [project page](https://github.com/social-manager-tools/twitter-bot.js/projects?query=is%3Aopen+sort%3Aname-asc).
* [v0.1.X](https://github.com/social-manager-tools/twitter-bot.js/projects/1)
* [v0.2.X](https://github.com/social-manager-tools/twitter-bot.js/projects/2)
* [v0.3.X](https://github.com/social-manager-tools/twitter-bot.js/projects/3)

## Sorry for snake_case
I love :snake: snake_case syntax sorry for this :sob: don't hate me.

[![](https://socialmanagertools.ptkdev.io/img/socialmanagertools_logo.png)](https://github.com/social-manager-tools)

# Social Manager Tools

[Social Manager Tools GUI](https://github.com/social-manager-tools/social-manager-tools)  
[InstagramBot.js](https://github.com/social-manager-tools/instagram-bot.js) ([LIB](https://github.com/social-manager-tools/instagram-bot-lib))  
[TwitterBot.js](https://github.com/social-manager-tools/twitter-bot.js) ([LIB](https://github.com/social-manager-tools/twitter-bot-lib))  
[FacebookBot.js](https://github.com/social-manager-tools/facebook-bot.js) ([LIB](https://github.com/social-manager-tools/facebook-bot-lib))  
[WordpressTelegramBot.js](https://github.com/social-manager-tools/wordpress-telegram-bot.js) ([LIB](https://github.com/social-manager-tools/wordpress-telegram-bot-lib))  
[MediumTelegramBot.js](https://github.com/social-manager-tools/medium-telegram-bot.js) ([LIB](https://github.com/social-manager-tools/medium-telegram-bot-lib))  

# License

GNU GENERAL PUBLIC LICENSE

Copyright (c) 2018 Patryk Rzucidło (PTKDev)


CHANGELOG.md
# v0.2.5 (October 02, 2018)
* Fix: Startup error


# v0.2.4 (September 26, 2018)
* Upgrade: puppeteer v1.6.2 


# v0.2.3 (September 15, 2018)
* Downgrade: puppeteer v1.4.0 


# v0.2.2 (September 15, 2018)
* Fix: RT the first tweet of reply (#6)
* Fix: RT/Like mode now work with words and hashtags
* Update: New user-agent
* Upgrade: puppeteer v1.8.0


# v0.2.0 (July 26, 2018)
* Fix: default path of logs in `config.js.tpl`
* Fix: Unliking already liked tweet
* Fix: Unretweeting already rtweeted tweet


# v0.1.0 (July 22, 2018)
* First Release.
* Feature: likemode_realistic
* Feature: rtmode_realistic


modules/common/utils.js
/**
 * Utils
 * =====================
 * Logger and other functions...
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <support@ptkdev.io> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1.0
 * @changelog:  0.1 initial release
 *
 */
require("colors");
class Utils {
    constructor(bot, config) {
        this.bot = bot;
        this.config = config;
        this.LOG_NAME = "utils";
        this.LOG = require("../logger/types");
        this.MAP_COLORS = require("../logger/types").MAP_COLORS;
        this.Log = require("../logger/Log");
        this.log = new this.Log(this.LOG_NAME, this.config);
    }

    /**
     * Donate
     * =====================
     * Patreon link
     *
     */
    donate() {
        this.log.warning("Bot work? Please donate for support this project!");
        this.log.warning("Donate with patreon: http://patreon.ptkdev.io");
        this.log.warning("Donate with paypal: http://paypal.ptkdev.io");
    }

    /**
     * Check updates
     * =====================
     * Bot is updated? Yes/no
     *
     */
    check_updates(version) {
        let request = require("request");
        let log = this.log;
        request.get("https://api.ptkdev.io/v1/bot/twitter/version/", function(err, res, last_release) {
            if (err) {
                log.error("Is impossible contact api.ptkdev.io server, wifi is on?");
            } else {
                if (version !== last_release) {
                    log.warning("Bot release v" + last_release + " available! Current version: v" + version);
                } else {
                    log.info("Bot is updated! :D");
                }
            }
        });
    }

    /**
     * Fix ui configs params
     * =====================
     * String aparms to booleans from social-manager-tools
     *
     */
    fixui(config) {
        config.bot_likeday_min = parseInt(config.bot_likeday_min);
        config.bot_likeday_max = parseInt(config.bot_likeday_max);
        if (config.chrome_headless == "enabled") {
            config.chrome_headless = false;
        } else {
            config.chrome_headless = true;
        }

        return config;
    }

    /**
     * Fix config
     * =====================
     * Fix previous and deprecated params
     *
     */
    fixconfig(config) {
        let max = 0;
        do {
            max++;
        } while ((60 / max * 24 * 20) > config.bot_likeday_min);
        config.bot_fastlike_max = max--;

        let min = 0;
        do {
            min++;
        } while ((60 / min * 24 * 20) > config.bot_likeday_max);
        config.bot_fastlike_min = min--;

        return config;
    }

    /**
     * Default config.js
     * =====================
     * Get default value if config.js is not updated from config.js.tpl
     *
     */
    default_config(config) {
        if (typeof config.debug === "undefined") {
            config.debug = true;
            this.log.error("config.debug use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.login === "undefined") {
            config.login = true;
            this.log.error("config.login use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.twitter_username === "undefined") {
            config.twitter_username = "ptkdev";
            this.log.error("config.twitter_username use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.twitter_password === "undefined") {
            config.twitter_password = "password";
            this.log.error("config.twitter_password use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.twitter_hashtag === "undefined") {
            config.twitter_hashtag = ["follow"];
            this.log.error("config.twitter_hashtag use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.twitter_pin === "undefined") {
            config.twitter_pin = "sms";
            this.log.error("config.twitter_pin use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_mode === "undefined") {
            config.bot_mode = "likemode_classic";
            this.log.error("config.bot_mode use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_likeday_min === "undefined") {
            config.bot_likeday_min = 300;
            this.log.error("config.bot_likeday_min use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_likeday_max === "undefined") {
            config.bot_likeday_max = 600;
            this.log.error("config.bot_likeday_max use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_sleep_night === "undefined" || (config.bot_sleep_night !== true && config.bot_sleep_night !== false)) {
            config.bot_sleep_night = true;
            this.log.error("config.bot_sleep_night use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_stop_sleep === "undefined") {
            config.bot_stop_sleep = "7:00";
            this.log.error("config.bot_stop_sleep use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_fastlike_min === "undefined") {
            config.bot_fastlike_min = 15;
            this.log.error("config.bot_fastlike_min use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_fastlike_max === "undefined") {
            config.bot_fastlike_max = 20;
            this.log.error("config.bot_fastlike_max use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.chrome_headless === "undefined") {
            config.chrome_headless = false;
            this.log.error("config.chrome_headless use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.chrome_options === "undefined") {
            config.chrome_options = ["--disable-gpu", "--no-sandbox", "--window-size=1920x1080"];
            this.log.error("config.chrome_options use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.executable_path === "undefined") {
            config.executable_path = "";
            this.log.error("config.executable_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.screenshot_path === "undefined") {
            config.screenshot_path = "./logs/screenshot/";
            this.log.error("config.screenshot_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.logerr_path === "undefined") {
            config.logerr_path = "./logs/errors.log";
            this.log.error("config.logerr_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.log_path === "undefined") {
            config.log_path = "./logs/debug.log";
            this.log.error("config.log_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.pin_path === "undefined") {
            config.pin_path = "./loginpin.txt";
            this.log.error("config.pin_path use the default value, update your config.js from config.js.tpl");
        }

        return config;
    }

    /**
     * Screenshot
     * =====================
     * Save screenshot from chrome
     *
     */
    async screenshot(func, name) {
        if (this.config.log.screenshot) {
            try {
                await this.bot.screenshot({ path: this.config.screenshot_path + this.config.twitter_username + "_" + func + "_" + name + ".jpg" });
                this.log.info("Cheese! Screenshot!");
            } catch (err) {
                this.log.error(this.LOG.WARNING, "screenshot", "error " + err);
            }
        }
    }

    /**
     * Random
     * =====================
     * Random number between two numbers
     *
     */
    random_interval(min, max) {
        return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
    }

    /**
     * Get random number between two numbers
     * @param min
     * @param max
     * @return {number}
     */
    random_number() {
        return (Math.floor(Math.random() * (20 - 10 + 1)) + 10);
    }

    /**
     * Mix array element
     * @param arr
     * @return array
     */
    mix_array(arr) {
        return arr.sort(function() {
            return 0.5 - Math.random(); 
        });
    }

    /**
     * Sleep
     * =====================
     * Zzz
     *
     */
    sleep(sec) {
        return new Promise(resolve => setTimeout(resolve, sec));
    }

    /**
     * Check is debug
     * @return {boolean}
     */
    is_debug() {
        return this.config.debug === true;
    }

    /**
     * Get random hash tag from config file
     * @return {string}
     */
    get_random_hash_tag() {
        return this.config.twitter_hashtag[Math.floor(Math.random() * this.config.twitter_hashtag.length)];
    }
}

module.exports = (bot, config, utils) => {
    return new Utils(bot, config, utils); 
};